File: tools/check_cutout.cs
#:project GateKit/GateKit.csproj
// Alpha-tested (cutout) shadow gate: a holed card must cast a holed shadow.
//
//   dotnet run tools/check_cutout.cs
//
// Renders the same holed card twice — once as BlendMode.Cutout, once as a plain opaque material —
// and compares.
//
// The card itself looks identical in both runs, which is the whole point of the bug. GL blending
// is enabled globally, so a texel with alpha 0 blends away in the colour pass whether or not the
// material is a cutout: the card LOOKS like a leaf either way. Only the depth pass differs, and
// without alpha testing it writes the full quad — so a leaf-shaped card casts a rectangular shadow.
//
// The measurement is therefore entirely about the shadow.
using GateKit;

const double MinShadowReduction = 0.40;   // the ring-with-gaps texture covers well under half the quad

var (cutPng, cutBlend, cutCutoff, cutCasts) = Run("cutout");
var (solPng, solBlend, solCutoff, solCasts) = Run("solid", "--solid");

Gate.Info($"cutout run: blend={cutBlend} cutoff={cutCutoff} castShadows={cutCasts}");
Gate.Info($"solid  run: blend={solBlend} cutoff={solCutoff} castShadows={solCasts}");
Gate.Blank();

var cut = GateImage.Load(cutPng);
var sol = GateImage.Load(solPng);

int sc = ShadowPixels(cut), ss = ShadowPixels(sol);
double reduction = ss > 0 ? 1.0 - (double)sc / ss : 0;
int darker = DarkerPixels(cut, sol);

Gate.Check("cutout material reports Cutout + casts", cutBlend == "Cutout" && cutCasts == "True",
           $"blend={cutBlend} castShadows={cutCasts}");
Gate.Check("shadow gains holes", reduction >= MinShadowReduction,
           $"{ss} -> {sc} shadow samples ({100 * reduction:F0}% less)");
Gate.Check("cutout never shadows MORE", darker == 0, $"{darker} px darker");

return Gate.Report("CUTOUT PASS — a holed card casts a holed shadow", "CUTOUT FAIL");

// Count neutral-grey pixels: the ground is warm-tinted, the shadow is neutral. Sampling every
// other pixel is plenty for an area measurement and keeps the gate quick.
static int ShadowPixels(GateImage im)
{
    int n = 0;
    for (int y = (int)(im.Height * 0.10); y < im.Height; y += 2)
    for (int x = 0; x < im.Width; x += 2)
    {
        var (r, g, b) = im[x, y];
        if (Math.Abs(r - g) < 12 && Math.Abs(g - b) < 12 && r < 190) n++;
    }
    return n;
}

// Pixels where the cutout run is materially darker than the solid one — a cutout must never
// shadow MORE than the shape it is cut from.
static int DarkerPixels(GateImage a, GateImage b)
{
    int n = 0;
    for (int y = (int)(a.Height * 0.10); y < a.Height; y += 3)
    for (int x = 0; x < a.Width; x += 3)
    {
        var (ar, ag, ab) = a[x, y];
        var (br, bg, bb) = b[x, y];
        if (ar + ag + ab < br + bg + bb - 25) n++;
    }
    return n;
}

static (string Png, string Blend, string Cutoff, string Casts) Run(string name, params string[] extra)
{
    string png = $"/tmp/cutgate_{name}.png";
    var args = new List<string> { "--shot", png, "--frames", "40" };
    args.AddRange(extra);
    string output = Gate.RunSketch("sketches/cutout_test.cs", args.ToArray());
    return (png,
            Gate.Text(output, @"blend=(\w+) cutoff=") ?? "?",
            Gate.Text(output, @"cutoff=([\d.]+)") ?? "?",
            Gate.Text(output, @"castShadows=(\w+)") ?? "?");
}

File: tools/check_cutout.py
#!/usr/bin/env python3
"""Alpha-tested (cutout) shadow gate: a holed card must cast a holed shadow.

    python3 tools/check_cutout.py

Renders the same holed card twice — once as BlendMode.Cutout, once as a plain opaque material —
and compares.

The card itself looks identical in both runs, which is the whole point of the bug. GL blending is
enabled globally, so a texel with alpha 0 blends away in the colour pass whether or not the
material is a cutout: the card LOOKS like a leaf either way. Only the depth pass differs, and
without alpha testing it writes the full quad — so a leaf-shaped card casts a rectangular shadow.

The measurement is therefore entirely about the shadow.
"""
import re
import subprocess
import sys
from pathlib import Path

from PIL import Image

ROOT = Path(__file__).resolve().parent.parent
MIN_SHADOW_REDUCTION = 0.40    # the ring-with-gaps texture covers well under half the quad


def run(name, flags):
    out = f"/tmp/cutgate_{name}.png"
    p = subprocess.run(["dotnet", "run", "sketches/cutout_test.cs", "--",
                        "--shot", out, "--frames", "40", *flags],
                       capture_output=True, text=True, cwd=ROOT)
    m = re.search(r"blend=(\w+) cutoff=([\d.]+) castShadows=(\w+)", p.stdout)
    if not m:
        print(f"  {name}: no report\n{p.stdout}\n{p.stderr}")
        sys.exit(1)
    return out, m.group(1), m.group(2), m.group(3)


def shadow_pixels(path):
    """Count neutral-grey pixels: the ground is warm-tinted, the shadow is neutral."""
    im = Image.open(path).convert("RGB")
    w, h = im.size
    n = 0
    for y in range(int(h * 0.10), h, 2):          # skip the stats overlay
        for x in range(0, w, 2):
            r, g, b = im.getpixel((x, y))
            if abs(r - g) < 12 and abs(g - b) < 12 and r < 190:
                n += 1
    return n


def darker_pixels(a_path, b_path):
    """Pixels where A is materially darker than B — a cutout must never shadow MORE."""
    a = Image.open(a_path).convert("RGB")
    b = Image.open(b_path).convert("RGB")
    w, h = a.size
    return sum(1 for y in range(int(h * 0.10), h, 3) for x in range(0, w, 3)
               if sum(a.getpixel((x, y))) < sum(b.getpixel((x, y))) - 25)


def main():
    cut_png, cut_blend, cut_cutoff, cut_casts = run("cutout", [])
    sol_png, sol_blend, sol_cutoff, sol_casts = run("solid", ["--solid"])
    print(f"  cutout run: blend={cut_blend} cutoff={cut_cutoff} castShadows={cut_casts}")
    print(f"  solid  run: blend={sol_blend} cutoff={sol_cutoff} castShadows={sol_casts}")

    sc, ss = shadow_pixels(cut_png), shadow_pixels(sol_png)
    reduction = 1 - sc / ss if ss else 0
    darker = darker_pixels(cut_png, sol_png)

    checks = [
        ("cutout material reports Cutout + casts", cut_blend == "Cutout" and cut_casts == "True",
         f"blend={cut_blend} castShadows={cut_casts}"),
        ("shadow gains holes", reduction >= MIN_SHADOW_REDUCTION,
         f"{ss} -> {sc} shadow samples ({100 * reduction:.0f}% less)"),
        ("cutout never shadows MORE", darker == 0, f"{darker} px darker"),
    ]

    print()
    ok = True
    for name, passed, detail in checks:
        print(f"  {'ok  ' if passed else 'FAIL'} {name:38s} {detail}")
        ok &= passed

    print()
    print("CUTOUT PASS — a holed card casts a holed shadow" if ok else "CUTOUT FAIL")
    sys.exit(0 if ok else 1)


if __name__ == "__main__":
    main()
